enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,546 @@
+controlled way for the next revision, or just keep them as the correct
+revision and reuse them.
+
+#### Post-reboot service
+
+We will add a new `post-reboot` service for approving the kubelet and
+the node Certificate Signing Requests. This service will also cleanup
+the bootstrap static pods resources when the OpenShift control plane
+is ready.
+
+Since we start with a liveCD, the bootstrap services (`bootkube`, `approve-csr`, etc.), `/etc` and `/opt/openshift` temporary files are written to the ephemeral filesystem of the live image, and not to the node's real filesystem.
+
+The files that we need to delete are under:
+
+* `/etc/kubernetes/bootstrap-secrets`
+* `/etc/kubernetes/bootstrap-configs`
+
+These files are required for the bootstrap control plane to start before it is replaced by the control plane operators.
+Once the OCP control plane static pods are deployed we can delete the files as they are no longer required.
+
+#### Prerequisites for a Single Node deployment with bootstrap-in-place
+The requirements are a subset of the requirements for user-provisioned infrastructure installation.
+1. Configure DHCP or set static IP addresses for the node.
+The node IP should be persistent, otherwise TLS SAN will be invalidated and will cause the communications between apiserver and etcd to fail.
+2. DNS records:
+* api.<cluster_name>.<base_domain>.
+* api-int.<cluster_name>.<base_domain>.
+* *.apps.<cluster_name>.<base_domain>.
+* <hostname>.<cluster_name>.<base_domain>.
+### Initial Proof-of-Concept
+
+A proof-of-concept implementation is available for experimenting with
+the design.
+
+To try it out: [bootstrap-in-place-poc](https://github.com/eranco74/bootstrap-in-place-poc.git)
+
+### Risks and Mitigations
+
+*What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.*
+
+*How will security be reviewed and by whom? How will UX be reviewed and by whom?*
+
+#### Custom Manifests for CRDs
+
+One limitation of single-node deployments not present in multi-node
+clusters is handling some custom resource definitions (CRDs). During
+bootstrapping of a multi-node cluster, the bootstrap host and real
+cluster hosts run in parallel for a time. This means that the
+bootstrap host can iterate publishing manifests to the API server
+until the operators running on the other hosts are up and define their
+CRDs. If it takes a little while for those operators to install their
+CRDs, the bootstrap host can wait and retry the operation. In a
+single-node deployment, the bootstrap environment and real node are
+not active at the same time. This means the bootstrap process may
+block if it tries to create a custom resource using a CRD that is not
+installed.
+
+While most CRDs are created by the `cluster-version-operator`, some
+CRDs are created later by the cluster operators. These CRDs from
+cluster operators are not present during bootstrapping:
```

Comment:
Is this just pointing out CRDs created later though no CRs are created at bootstrap? Or are we creating CRs and thus have a cycle/problem? What is being done to resolve this?

Reply:
It's just pointing out that some CRDs are generated later. There isn't currently a dependency issue in the installer itself, but if the user adds extra manifests using these CRDs there could be.

I guess ideally all CRDs would be installed by the CVO, but it's not clear why some are and some are not.